libcontainer/nsenter/nsexec.c

```diff
@@ -56,166 +54,248 @@ int setns(int fd, int nstype)
+	if (pipenum == -1) {
 		return;
 	}
-	if (strcmp(val, "setns") != 0) {
-		pr_perror("Invalid inittype %s", val);
-		exit(1);
-	}
 
-	val = getenv("_LIBCONTAINER_INITPIPE");
-	if (val == NULL) {
-		pr_perror("Child pipe not found");
-		exit(1);
-	}
-	pipenum = atoi(val);
-	snprintf(buf, sizeof(buf), "%d", pipenum);
-	if (strcmp(val, buf)) {
-		pr_perror("Unable to parse _LIBCONTAINER_INITPIPE");
-		exit(1);
-	}
+	// Retrieve the netlink header
+	struct nlmsghdr	nl_msg_hdr;
+	int				len;
 
-	char nlbuf[NLMSG_HDRLEN];
-	struct nlmsghdr *nh;
-	if ((n = read(pipenum, nlbuf, NLMSG_HDRLEN)) != NLMSG_HDRLEN) {
-		pr_perror("Failed to read netlink header, got %d", n);
+	if ((len = read(pipenum, &nl_msg_hdr, NLMSG_HDRLEN)) != NLMSG_HDRLEN) {
+		pr_perror("Failed to read netlink header, got %d instead of %d",
+				  len, NLMSG_HDRLEN);
 		exit(1);
 	}
 
-	nh = (struct nlmsghdr *)nlbuf;
-	if (nh->nlmsg_type == NLMSG_ERROR) {
-		pr_perror("Invalid netlink header message");
+	if (nl_msg_hdr.nlmsg_type == NLMSG_ERROR) {
+		pr_perror("failed to read netlink message");
 		exit(1);
 	}
-	if (nh->nlmsg_type != INIT_MSG) {
-		pr_perror("Unexpected netlink message type %d", nh->nlmsg_type);
+	if (nl_msg_hdr.nlmsg_type != INIT_MSG) {
+		pr_perror("unexpected msg type %d", nl_msg_hdr.nlmsg_type);
 		exit(1);
 	}
-	// read the netlink payload
-	len = NLMSG_PAYLOAD(nh, 0);
-	char data[len];
-	if ((n = read(pipenum, data, len)) != len) {
-		pr_perror("Failed to read netlink payload, got %d", n);
+
+	// Retrieve data
+	int		nl_total_size = NLMSG_PAYLOAD(&nl_msg_hdr, 0);
+	char	data[nl_total_size];
+
+	if ((len = read(pipenum, data, nl_total_size)) != nl_total_size) {
+		pr_perror("Failed to read netlink payload, got %d instead of %d",
+				  len, nl_total_size);
 		exit(1);
 	}
 
-	int start = 0;
-	struct nlattr *attr;
-	while (start < len) {
-		int payload_len;
-		attr = (struct nlattr *)((void *)data + start);
-		start += NLA_HDRLEN;
-		payload_len = attr->nla_len - NLA_HDRLEN;
-		switch (attr->nla_type) {
-		case PID_ATTR:
-			pid = (pid_t) readint32(data + start);
-			break;
-		case CONSOLE_PATH_ATTR:
-			consolefd = open((char *)data + start, O_RDWR);
+	// Process the passed attributes
+	int				start = 0;
+	uint32_t		cloneflags = -1;
+	uint8_t			is_setgroup = 0;
+	int				consolefd = -1;
+	int				uidmap_start = -1, uidmap_len = -1;
+	int				gidmap_start = -1, gidmap_len = -1;
+	int				payload_len;
+	struct nlattr	*nlattr;
+
+	while (start < nl_total_size) {
+		nlattr		= (struct nlattr *)(data + start);
+		start		+= NLA_HDRLEN;
+		payload_len = nlattr->nla_len - NLA_HDRLEN;
+
+		if (nlattr->nla_type == CLONE_FLAGS_ATTR) {
+			cloneflags = readint32(data + start);
+		} else if (nlattr->nla_type == CONSOLE_PATH_ATTR) {
+			// get the console path before setns because it may change mnt namespace
+			char console[payload_len + 1];
+
+			strncpy(console, data + start, payload_len);
+			console[payload_len] = '\0';
+			consolefd = open(console, O_RDWR);
```

Comment:
we dont need to copy here, just open the consolefd directly like the previous code `consolefd = open((char *)data + start, O_RDWR);`